src/store/styleguide/typescript-react-analyzer/index.ts

```diff
@@ -0,0 +1,114 @@
+import { getReactComponentExports, ReactComponentExport } from './typescript/react';
+import { PatternFileInfo, ReactPattern } from './react-pattern';
+import { StyleguideAnalyzer } from '../styleguide-analyzer';
+import * as ts from 'typescript';
+
+import * as FileUtils from 'fs';
+import * as PathUtils from 'path';
+import { PatternType } from '../../pattern/pattern-type';
+
+export class TypescriptReactAnalyzer extends StyleguideAnalyzer<ReactPattern> {
+	protected program: ts.Program;
+
+	public getPatternType(): PatternType {
+		return PatternType.react;
+	}
+
+	public analyze(path: string): ReactPattern[] {
+		const patterns: ReactPattern[] = [];
+
+		const fileInfos = walkDirectoriesAndCollectPatterns(path);
+
+		const declarationFiles = fileInfos.map(patternInfo => patternInfo.declarationFilePath);
+
+		this.program = ts.createProgram(declarationFiles, {}, undefined, this.program);
+
+		fileInfos.forEach(fileInfo => {
+			const sourceFile = this.program.getSourceFile(fileInfo.declarationFilePath);
+			const exports = getReactComponentExports(sourceFile, this.program);
+
+			exports.forEach(exportInfo => {
+				const id = generatePatternId(path, fileInfo, exportInfo);
+				const name = getPatternName(fileInfo, exportInfo);
+
+				const pattern = new ReactPattern({
+					id,
+					name,
+					analyzer: this,
+					fileInfo,
+					exportInfo
+				});
+
+				patterns.push(pattern);
+			});
+		});
+
+		return patterns;
+	}
+}
+
+function getPatternName(fileInfo: PatternFileInfo, exportInfo: ReactComponentExport): string {
+	const baseName = PathUtils.basename(fileInfo.jsFilePath, '.js');
+	const directoryName = PathUtils.basename(fileInfo.directory);
+	const name = exportInfo.name ? exportInfo.name : baseName !== 'index' ? baseName : directoryName;
```

Comment:
This would make it a bit more readable:
`exportInfo.name || baseName !== 'index' ? baseName : directoryName`